torch_geometric/explain/algorithm/gnn_explainer.py

```diff
@@ -0,0 +1,546 @@
+            (default: :obj:`100`)
+        lr (float, optional): The learning rate to apply.
+            (default: :obj:`0.01`)
+        shared_mask (bool, optional): Whether to share the mask for node
+            features. When set to :obj:`True`, the node feature mask will be
+            shared across all nodes. (default: :obj:`True`). Only used when the
+            node mask type is :obj:`MaskType.attributes`.
+        **kwargs (optional): Additional hyper-parameters to override default
+            settings in
+            :attr:`~torch_geometric.explain.algorithm.GNNExplainer.coeffs`.
+    """
+
+    coeffs = {
+        'edge_size': 0.005,
+        'edge_reduction': 'sum',
+        'node_feat_size': 1.0,
+        'node_feat_reduction': 'mean',
+        'edge_ent': 1.0,
+        'node_feat_ent': 0.1,
+        'EPS': 1e-15,
+    }
+
+    def __init__(
+        self,
+        epochs: int = 100,
+        lr: float = 0.01,
+        **kwargs,
+    ):
+        super().__init__()
+        self.epochs = epochs
+        self.lr = lr
+        self.coeffs.update(kwargs)
+        # node_mask will never actually be None
+        self.node_mask = None
+        # edge_mask is only None if edge_mask_type is != MaskType.object
+        self.edge_mask = None
+
+    def forward(
+        self,
+        model: torch.nn.Module,
+        x: Tensor,
+        edge_index: Tensor,
+        explainer_config: ExplainerConfig,
+        model_config: ModelConfig,
+        target: Tensor,
+        target_index: Optional[Union[int, Tensor]] = None,
+        node_index: Optional[int] = None,
+        **kwargs,
+    ) -> Explanation:
+
+        if isinstance(target_index, Tensor) and target_index.numel() > 1:
+            raise ValueError(
+                "GNNExplainer only supports single target index for now")
+
+        assert model_config.task_level in [
+            ModelTaskLevel.graph, ModelTaskLevel.node
+        ]
+
+        model.eval()
+
+        if model_config.task_level == ModelTaskLevel.node:
+            node_mask, edge_mask = self._explain_node(model, x, edge_index,
+                                                      explainer_config,
+                                                      model_config, target,
+                                                      node_index, target_index,
+                                                      **kwargs)
+        elif model_config.task_level == ModelTaskLevel.graph:
+            node_mask, edge_mask = self._explain_graph(model, x, edge_index,
+                                                       explainer_config,
+                                                       model_config, target,
+                                                       target_index, **kwargs)
+
+        if explainer_config.node_mask_type == MaskType.object:
+            node_feat_mask = None
+        else:
+            node_feat_mask = node_mask
+            node_mask = None
```

Comment:
More of a broader point, but what is the purpose of separating node_mask vs node_feat_mask in the explanation class?

node_mask is of shape [N], while node_feat_mask is of shape [N, F] or [1, F]. Given the dimension is already variable, perhaps we could just combine them? Same for edge_mask/edge_feat_mask

Also, is the docstring in explanations.py outdated?
`node_feat_mask (Tensor, optional): Node-level feature mask with shape [num_nodes, num_node_features]`

Reply:
So for now the node_feat_mask is always [N,F] (see comment above).

Reply:
Ah makes sense now.